src/material-examples/input-error-state-matcher/input-error-state-matcher-example.ts

```diff
@@ -0,0 +1,26 @@
+import {Component} from '@angular/core';
+import {FormControl, FormGroupDirective, NgForm, Validators} from '@angular/forms';
+import {ErrorStateMatcher} from '@angular/material/core';
+
+/** Error when invalid control is dirty, touched, or submitted. */
+export class MyErrorStateMatcher implements ErrorStateMatcher {
+  isErrorState(control: FormControl | null, form: FormGroupDirective | NgForm | null): boolean {
+    const isSubmitted = form && form.submitted;
+    return !!(control && control.invalid && (control.dirty || control.touched || isSubmitted));
+  }
+}
+
+/** @title Input with a custom ErrorStateMatcher */
+@Component({
+  selector: 'input-error-state-matcher-example',
+  templateUrl: './input-error-state-matcher-example.html',
+  styleUrls: ['./input-error-state-matcher-example.css'],
+})
+export class InputErrorStateMatcherExample {
+  emailFormControl = new FormControl('', [
+    Validators.required,
+    Validators.email,
+  ]);
+
+  matcher = new MyErrorStateMatcher();
```

Comment:
In my experience, the custom matchers are very practical for showing errors when a parent FormGroup is invalid. As in,

* Password/confirm must be equal
* `end` date must be after `begin` date.

This requires either checking `control.parent.valid` or [using a class property in the method](https://github.com/angular/material2/blob/7fe1b81a47b6de036ae85c183328400e45c40b15/src/lib/input/input.spec.ts#L1298-L1302). Maybe consider expanding the example? Or I'd be happy to give it a shot in a followup.

Reply:
Yeah that sounds like a good thing to add. I'd like to keep this example around too since it's a little less complex and we don't have to worry about also making a custom validator